From faa323ccca03b0c00a1afab208b7faf6a127567a Mon Sep 17 00:00:00 2001
From: saikumaru <saikumaru@ami.com>
Date: Tue, 9 Sep 2025 14:46:44 +0530
Subject: [PATCH] Add a clear fault flag and implement support to execute clear
  fault when ltc4286 STATUS_WORD != 0

Modifications introduced in this commit:
1. Add a new clear-fault flag.
2. When the DTS includes enable-clear-fault-flag, the clear-fault flag is set to true.
3. Execute the clear-fault operation when the STATUS_WORD of ltc4286 is not equal to 0.

Signed-off-by: saikumaru <saikumaru@ami.com>
Upstream-Status: Backport [Nuvoton SDK/ASPEED SDK migration (git commit cb052d637)]
---
 drivers/hwmon/pmbus/ltc4286.c    | 8 ++++++++
 drivers/hwmon/pmbus/pmbus.h      | 3 +++
 drivers/hwmon/pmbus/pmbus_core.c | 6 +++++-
 3 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/drivers/hwmon/pmbus/ltc4286.c b/drivers/hwmon/pmbus/ltc4286.c
index 8715d380784a..4551c3ba9e9b 100644
--- a/drivers/hwmon/pmbus/ltc4286.c
+++ b/drivers/hwmon/pmbus/ltc4286.c
@@ -150,6 +150,14 @@ static int ltc4286_probe(struct i2c_client *client)
 
 	info->m[PSC_CURRENT_OUT] = 1024 * rsense;
 
+	/* Setup clear fault flag */
+	if (device_property_read_bool(dev, "enable-clear-fault-flag")) {
+		info->clear_fault_flag = true;
+	} else {
+		info->clear_fault_flag = false;
+	}
+
+
 	return pmbus_do_probe(client, info);
 }
 
diff --git a/drivers/hwmon/pmbus/pmbus.h b/drivers/hwmon/pmbus/pmbus.h
index e499cdae9442..1f52b5df2a6a 100644
--- a/drivers/hwmon/pmbus/pmbus.h
+++ b/drivers/hwmon/pmbus/pmbus.h
@@ -474,6 +474,9 @@ struct pmbus_driver_info {
 	/* custom attributes */
 	const struct attribute_group **groups;
 
+	/* Decide whether to use clear fault when STATUS_WORD(0x79) != 0. */
+	bool clear_fault_flag;
+
 	/*
 	 * Some chips need a little delay between SMBus communication. When
 	 * set, the generic PMBus helper functions will wait if necessary
diff --git a/drivers/hwmon/pmbus/pmbus_core.c b/drivers/hwmon/pmbus/pmbus_core.c
index 7150f12d2630..05729514a269 100644
--- a/drivers/hwmon/pmbus/pmbus_core.c
+++ b/drivers/hwmon/pmbus/pmbus_core.c
@@ -707,8 +707,12 @@ static int pmbus_get_status(struct i2c_client *client, int page, int reg)
 		status = _pmbus_read_byte_data(client, page, reg);
 		break;
 	}
-	if (status < 0)
+	
+	if (data->info->clear_fault_flag && status != 0)
+		pmbus_clear_faults(client);
+	else if (status < 0)
 		pmbus_clear_faults(client);
+
 	return status;
 }
 
-- 
2.34.1

